mono/utils/mono-rand-windows.c

```diff
@@ -7,63 +7,33 @@
 */
 #include <config.h>
 #include <glib.h>
+#if defined(HOST_WIN32) && (G_HAVE_API_SUPPORT(HAVE_CLASSIC_WINAPI_SUPPORT) || G_HAVE_API_SUPPORT(HAVE_UWP_WINAPI_SUPPORT))
 #include "mono-error.h"
 #include "mono-error-internals.h"
 #include "mono-rand.h"
-
-#if defined(HOST_WIN32)
+#include "mono-rand-windows-internals.h"
 #include <windows.h>
-#include "mono/utils/mono-rand-windows-internals.h"
+#include <bcrypt.h>
+#include <limits.h>
 
-#if G_HAVE_API_SUPPORT(HAVE_CLASSIC_WINAPI_SUPPORT)
-#ifndef PROV_INTEL_SEC
-#define PROV_INTEL_SEC		22
-#endif
-#ifndef CRYPT_VERIFY_CONTEXT
-#define CRYPT_VERIFY_CONTEXT	0xF0000000
+#ifndef BCRYPT_USE_SYSTEM_PREFERRED_RNG
+#define BCRYPT_USE_SYSTEM_PREFERRED_RNG (2)
 #endif
 
-MONO_WIN32_CRYPT_PROVIDER_HANDLE
-mono_rand_win_open_provider (void)
+long
+mono_rand_win_gen (guchar *buffer, size_t buffer_size)
 {
-	MONO_WIN32_CRYPT_PROVIDER_HANDLE provider = 0;
-
-	/* There is no need to create a container for just random data,
-	 * so we can use CRYPT_VERIFY_CONTEXT (one call) see:
-	 * http://blogs.msdn.com/dangriff/archive/2003/11/19/51709.aspx */
-
-	/* We first try to use the Intel PIII RNG if drivers are present */
-	if (!CryptAcquireContext (&provider, NULL, NULL, PROV_INTEL_SEC, CRYPT_VERIFY_CONTEXT)) {
-		/* not a PIII or no drivers available, use default RSA CSP */
-		if (!CryptAcquireContext (&provider, NULL, NULL, PROV_RSA_FULL, CRYPT_VERIFY_CONTEXT)) {
-			/* exception will be thrown in managed code */
-			provider = 0;
-		}
+	while (buffer_size > 0) {
+		ULONG const size = (ULONG)MIN (buffer_size, ULONG_MAX);
+		long const status = BCryptGenRandom (NULL, buffer, size, BCRYPT_USE_SYSTEM_PREFERRED_RNG);
```

Comment:
Will this take us up to Windows 7 when using BCryptGenRandom and BCRYPT_USE_SYSTEM_PREFERRED_RNG? Looking at the MSDN documentation it says:

> Windows Vista:  This flag is not supported.

I believe Mono should still supporting Vista and WINVER and _WIN32_WINNT is set to 0x0600 so we need to double check this since Unity might still have dependencies on Vista and forward.

Reply:
Missed the last line in the PR description, so taking this to Win7 might not be possible unless we check with Unity. Another alternative would be to take the implementation currently under uwp as a start, should work under Vista.

Reply:
https://unity3d.com/unity/system-requirements

Reply:
Jupp, that's why we are still on Vista with mono.

Reply:
But we could probably still retire the old CAPI implementation and use the implementation from the UWP source file?

Reply:
Agreed I'll take the UWP version. It is pretty close to this. Keep Vista going as Unity requires.